crates/fluvio-version-manager/src/command/install.rs

```diff
@@ -0,0 +1,285 @@
+//! Install Command
+//!
+//! Downloads and stores the sepecific Fluvio Version binaries in the local
+//! FVM cache.
+
+use std::fs::{File, create_dir, copy, write};
+use std::io::{Cursor, Read};
+use std::path::PathBuf;
+
+use anyhow::{Error, Result};
+use clap::Parser;
+use colored::Colorize;
+use serde::{Deserialize, Serialize};
+use semver::Version;
+use sha2::{Digest, Sha256};
+use surf::StatusCode;
+use tempfile::TempDir;
+use url::Url;
+
+use fluvio_hub_util::HUB_REMOTE;
+use fluvio_hub_util::fvm::{Client, PackageSet, DEFAULT_PKGSET, Channel};
+
+use crate::GlobalOptions;
+use crate::common::TARGET;
+use crate::common::fluvio::{fluvio_bin_path, create_fluvio_dir};
+use crate::common::notify::Notify;
+use crate::common::workdir::fvm_versions_path;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct VersionManifest {
+    pub channel: Channel,
+    pub version: Version,
+}
+
+/// The name printed to CLI in place of `default` for the Package Set
+const DEFAULT_PACKAGE_SET_DISPLAY_NAME: &str = "fluvio";
+
+/// The name of the manifest file for the Package Set
+const PACKAGE_SET_MANIFEST_FILENAME: &str = "manifest.json";
+
+/// The `install` command is responsible of installing the desired Package Set
+#[derive(Debug, Parser)]
+pub struct InstallOpt {
+    #[command(flatten)]
+    global_opts: GlobalOptions,
+    /// Version to install
+    #[arg(index = 1, default_value_t = Channel::Stable)]
+    version: Channel,
+    /// Registry used to fetch Fluvio Versions
+    #[arg(long, env = "HUB_REGISTRY_URL", default_value = HUB_REMOTE)]
+    registry: Url,
+}
+
+impl InstallOpt {
+    pub async fn process(&self) -> Result<()> {
+        if !fluvio_bin_path()?.exists() {
+            create_fluvio_dir()?;
+        }
+
+        self.install_package().await?;
+        Ok(())
+    }
+
+    async fn install_package(&self) -> Result<()> {
+        let pkgset = self.fetch_pkgset().await?;
+        let version_path = fvm_versions_path()?.join(self.version.to_string());
+
+        if version_path.exists() {
+            self.notify_warn(format!(
+                "Version {} is already installed. Aborting installation.",
+                self.version.to_string().bold()
+            ));
+
+            if matches!(self.version, Channel::Stable) || matches!(self.version, Channel::Latest) {
+                self.notify_help(format!(
+                    "Use {} to update the local installation for this channel",
+                    "fvm update".bold()
+                ));
+            }
+
+            return Ok(());
+        }
+
+        // The `tmp_dir` must be dropped after copying the binaries to the
+        // destination directory. By dropping `tmp_dir` the directory will be
+        // deleted from the filesystem.
+        let tmp_dir = TempDir::new()?;
+
+        self.download_artifacts(&tmp_dir, &pkgset).await?;
```

Comment:
Theres some special behavior going on here, the `TempDir` represents the temporal directory similar to a Database ORM's active model.

If I [drop `TempDir`, then the directory is deleted](https://docs.rs/tempdir/latest/tempdir/struct.TempDir.html#method.new), so by having the `TempDir` owned by the `download_artifacts` it would have to copy the downloaded binaries to a destination directory before returning.

To keep it simple/explicit, I rather specifying the out dir and handle it here.

Reply:
That's fine, but downloading/caching logic should be isolated from the installer.  All the installer needs to do is check and ask the downloader to see if pkg exists.  The downloader will return if it is cached; otherwise, it will download from the hub.   It should be probably moved to hub util.